Implement initial Assets API client with CLI support

eduardiazf opened the pull request:
## Summary
This PR implements a complete Assets API client for the Frequenz platform, including a command-line interface for interacting with microgrid assets. The implementation provides a robust foundation for asset management operations with comprehensive error handling and testing.

## Changes Made

### 🚀 New Features
- **Assets API Client** (`_client.py`): Complete gRPC client implementation for the Assets API
- **CLI Interface** (`cli/__main__.py`): Command-line tool with `microgrid` command for retrieving microgrid details
- **Type System** (`types.py`): Data classes for `Microgrid`, `DeliveryArea`, and `Location` with protobuf integration
- **Exception Handling** (`exceptions.py`): Custom exception hierarchy with JSON serialization support

### 🔧 Dependencies & Configuration
- Added new core dependencies: `frequenz-api-assets`, `frequenz-api-common`, `frequenz-client-base`, `grpcio`
- Introduced optional `cli` dependency group with `asyncclick`
- Updated `pyproject.toml` to include CLI dependencies

### 🧪 Testing Infrastructure
- Comprehensive test suite with pytest fixtures and helpers
- Mock factories for creating test objects
- Assertion helpers for common validation patterns
- Test coverage for various microgrid field combinations and error scenarios

## Technical Details

### Client Architecture
- Extends `BaseApiClient` for consistent authentication and connection management

### CLI Features
- Async command execution with proper asyncio event loop management
- Environment variable support for API credentials (`ASSETS_API_AUTH_KEY`, `ASSETS_API_SIGN_SECRET`)
- JSON output formatting for easy integration with other tools

## Usage Examples

```bash
# Get microgrid details
assests-cli microgrid 123

# With custom server and credentials
asssets-cli --server-url grpc://api.example.com:443 \
  --auth-key your-key --sign-secret your-secret microgrid 123
```

## Testing
- Comprehensive test coverage for success and error scenarios
- Mock-based testing with proper cleanup and isolation
- Parametrized tests for different field combinations

## Breaking Changes
- Updated dependency structure with new optional groups

## Impact
This implementation provides a solid foundation for asset management operations and establishes patterns that can be extended for other asset types in the future. The CLI interface makes it easy for users to interact with the Assets API directly from the command line, while the comprehensive client library enables programmatic access for integration into larger applications.

eduardiazf committed:
docs: update RELEASE_NOTES.md for Assets API client

- Document new features, breaking changes, and dependencies
- Add upgrade instructions and CLI usage details

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
Update dependencies in pyproject.toml and clean up assets client code

- Added new dependencies: frequenz-api-assets, frequenz-api-common, frequenz-client-base, and grpcio.
- Introduced optional dependency group 'cli' with asyncclick.
- Updated existing dependencies for frequenz-client-assets to include 'cli'.
- Removed deprecated 'delete_me' function from assets client and updated module docstring.

Updates dependencies and cleans up assets client

Updates project dependencies to integrate with new APIs and client libraries.

Introduces a 'cli' optional dependency group, including `asyncclick`.

Removes deprecated function from assets client for a cleaner API.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax commented on pyproject.toml:
```diff
@@ -27,6 +27,10 @@ classifiers = [
 dependencies = [
   "typing-extensions >= 4.12.2, < 5",
+  "frequenz-api-assets @ git+https://github.com/frequenz-floss/frequenz-api-assets.git@v0.x.x",
+  "frequenz-api-common >= 0.8.0, < 0.9.0",
```
We are committing to make v0.x API released backwards compatible, so you can use `frequenz-api-common >= 0.8.0, < 1` here, so we don't need to bump this every time a new common api version is released.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -0,0 +1,122 @@
+        )
+
+    @property
+    def stub(self) -> PlatformAssetsStub:
```
This should be `PlatformAssetsAsyncStub` (note the `Async`). You can find an explanation in the [note in the `BaseApiClient` docs](https://frequenz-floss.github.io/frequenz-client-base-python/latest/reference/frequenz/client/base/client/#frequenz.client.base.client.BaseApiClient).

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -0,0 +1,122 @@
+            response = await cast(
+                Awaitable[PBGetMicrogridResponse],
+                self.stub.GetMicrogrid(request),
+            )
```
When using the `AsyncStub`, this cast should not be necessary anymore.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -0,0 +1,122 @@
+            )
+
+            if response.microgrid is None:
+                raise NotFoundError(microgrid_id)
```
Is this correct? Shouldn't the API return a NOT FOUND error instead of a "valid" response with a "missing" microgrid field? Also, I don't think you'll ever get a `None` in here, if the field is missing, the default will be used. To check if a protobuf field is missing you need to use `response.HasField("microgrid")` (yes, it sucks). You can get more context in the [protobuf docs about field presence](https://protobuf.dev/programming-guides/field_presence/).

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -0,0 +1,122 @@
+            if e.code() == StatusCode.UNAUTHENTICATED:
+                raise AuthenticationError(e.details()) from e
+            if e.code() == StatusCode.UNAVAILABLE:
+                raise ServiceUnavailableError(e.details()) from e
```
We already have a tool to do all these exception conversions: [`frequenz.client.base.client.call_stub_method`](https://frequenz-floss.github.io/frequenz-client-base-python/latest/reference/frequenz/client/base/client/#frequenz.client.base.client.call_stub_method).

llucax commented on src/frequenz/client/assets/exceptions.py:
And we have all these exceptions wrappers too: [`frequenz.client.base.exception`](https://frequenz-floss.github.io/frequenz-client-base-python/latest/reference/frequenz/client/base/exception/#frequenz.client.base.exception). For example [in the microgrid API we just re-export all of them](https://github.com/frequenz-floss/frequenz-client-microgrid-python/blob/v0.x.x/src/frequenz/client/microgrid/_exception.py). Mainly because the intention was that users don't need to depend directly on `frequenz-client-base`, which is more like an internal library. Now I have my doubts, as it makes it more obscure that actually all the exception classes in all clients are actually the same class, which is important for `try/catch`, as if one thinks they are different could attempt to catch exception for different clients just using the exception class in the `except`, and that will not work. But this is probably a topic for another time :D

llucax commented on src/frequenz/client/assets/types.py:
This will soon go into `frequenz-client-common`, but just a heads up, for now we can keep a copy here. There are a few things there you can already use, and probably should, I will put them in separate comments).

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,171 @@
+    """
+
+    id: int
+    enterprise_id: int
```
You should use the `Id` classes in common:

* [`frequenz.client.common.microgrid.EnterpriseId`](https://frequenz-floss.github.io/frequenz-client-common-python/latest/reference/frequenz/client/common/microgrid/#frequenz.client.common.microgrid.EnterpriseId)
* [`frequenz.client.common.microgrid.MicrogridId`](https://frequenz-floss.github.io/frequenz-client-common-python/latest/reference/frequenz/client/common/microgrid/#frequenz.client.common.microgrid.MicrogridId)

llucax commented on src/frequenz/client/assets/__main__.py:
```diff
@@ -0,0 +1,190 @@
+
+import asyncio
+import json
+from typing import Optional
```
coding style: We don't use `Optional[T]`, we use `T | None` instead.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -0,0 +1,122 @@
+This module provides a client for the Assets API.
+"""
+
+from typing import Awaitable, Optional, cast
```
Importing `Awaitable` (and any container class) from `typing` is deprecated, you should import form `collections.abc` when possible.

llucax commented on tests/conftest.py:
```diff
@@ -0,0 +1,81 @@
+
+
+@pytest.fixture
+def client_setup() -> Generator[ClientSetup, None, None]:
```
This can be simplified using `Iterator[ClientSetup]` instead of `Generator`.

llucax commented on tests/conftest.py:
```diff
@@ -0,0 +1,81 @@
+from unittest.mock import AsyncMock
+
+import pytest
+from typing_extensions import Generator
```
Import from `collections.abc`. Use `typing_extensions` only when the type you want to use is not present in the minimum supported Python version.

llucax commented on tests/conftest.py:
```diff
@@ -0,0 +1,81 @@
+import pytest
+from typing_extensions import Generator
+
+from frequenz.client.assets import AssetsApiClient as Client
```
I recommend not using aliases unless really necessary, it makes the code more obscure, as to know what are you referring to you need to go to the import line.

llucax commented on tests/helpers/assertions.py:
```diff
@@ -0,0 +1,53 @@
+
+
+class AssertionHelpers:
+    """Helper methods for common assertions."""
```
coding style: Why grouping this in a class? We try to avoid unnecessary nesting. I would just make all these static methods global functions.

llucax commented on tests/helpers/factories.py:
```diff
@@ -0,0 +1,66 @@
+    code_type: str = MockData.delivery_area_type,
+) -> AsyncMock:
+    """Create delivery area mocks."""
+    mock = AsyncMock()
```
It might be useful to add a name and make it mimic the actual delivery area class:

`AsyncMock(name="delivery_area_mock", spec=DeliveryArea)`

With this, if anyone tries to use anything in the mock that is not in the `DeliveryArea` class, it will raise an error instead of silently create a new mock for it.

llucax commented on RELEASE_NOTES.md:
```diff
@@ -2,16 +2,44 @@
-<!-- Here goes notes on how to upgrade from previous versions, including deprecations and what they should be replaced with -->
+**Breaking Changes:**
+
+- Removed deprecated `delete_me` function
```
This function was never intended to be used, it is just part of the template we use to create repositories, so I wouldn't list it in the release notes, nor in the PR description.

llucax replied:
This one is still pending :)

llucax reviewed (commented):
The PR looks good to me in general, although I made quite a few comments, mostly to point to some common tools you might not know about or point out some coding style issues, so everything quite minor. Nevertheless it would be nice to address them to keep the coding style consistent and avoid code duplication.

In general I would also recommend to split into more commits for next time, to make reviewing easier. For example you could separate updating dependencies, adding the client class, adding the CLI interface, etc.

It is usually much less demanding, at least when reviewing commit-by-commit, to focus on only one set of small changes than having many unrelated changes in mind while going through diffs and files.

eduardiazf committed:
chore(deps): Change max version to 1 for frequenz-api-common

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
feat(client): Enhance Assets API client and update dependencies

- Added `frequenz-client-common` as a new dependency in `pyproject.toml`.
- Refactored error handling in the Assets API client to use `ApiClientError` instead of `AssetsApiError`.
- Updated type hints for `auth_key` and `sign_secret` to support `str | None`.
- Improved exception handling in `get_microgrid_details` method.
- Updated `Microgrid` class to use `MicrogridId` and `EnterpriseId` types for better type safety.

These changes improve the robustness and clarity of the Assets API client while ensuring compatibility with the latest dependencies.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
fix: Update typing-extensions dependency and refactor assertions

- Updated `typing-extensions` version in `pyproject.toml` to `>= 4.13.0, < 5`.
- Refactored assertions in test files to remove the `AssertionHelpers` class, replacing static methods with direct function calls for clarity and simplicity.
- Enhanced the `create_*_mock` functions in factories to include specific names and specs for better debugging.

These changes improve dependency management and streamline test assertions for better readability.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
refactor(client): Update Assets API client and exceptions

- Added `from __future__ import annotations` to improve type hinting.
- Changed the return type of the `stub` property to `PlatformAssetsAsyncStub` for better async support.
- Introduced new exceptions `DataLoss` and `EntityAlreadyExists` in the exceptions module.
- Updated the `Microgrid` class to ensure integer conversion for `id` and `enterprise_id`.

These changes enhance type safety and improve the overall structure of the Assets API client.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
fix(client): Add type ignore for async stub casting in Assets API client

- Added a type ignore comment in the `stub` property to handle the casting from a sync stub to an async stub.
- This change ensures compatibility with the async implementation while maintaining type safety.

These modifications improve the functionality of the Assets API client by addressing type issues rel

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
fix(client): Update exception reference in Assets API client documentation

- Corrected the reference to the `GrpcError` class in the docstring of the `get_microgrid_details` method to point to the appropriate module.
- This change improves the accuracy of the documentation and helps users understand the error handling better.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax commented:
It looks like commit [`6902612` (#21)](https://github.com/frequenz-floss/frequenz-client-assets-python/pull/21/commits/690261201c8233595ae50933111dc14b595fca6d) is missing a title.

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,174 @@
+
+from dataclasses import dataclass
+from datetime import datetime, timezone
+from typing import Any, Optional
```
Nitpick: There is still one `Optional` here :)

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,174 @@
+        """
+        return DeliveryArea(code=pb.code, code_type=str(pb.code_type))
+
+    def to_dict(self) -> dict[str, Any]:
+        """
+        Convert the DeliveryArea instance to a dictionary.
+
+        Returns:
+            A dictionary containing the delivery area details.
+        """
+        return {"code": self.code, "code_type": self.code_type}
```
These `to_dict()` are not necessary, you can use [`dataclasses.asdict()`](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) on any `dataclass`.

llucax commented on tests/conftest.py:
```diff
@@ -0,0 +1,81 @@
+        mock_response=mock_response,
+    )
+
+    # Cleanup: reset ALL mocks automatically
+    mock_stub.reset_mock()
+    mock_microgrid.reset_mock()
+    mock_delivery_area.reset_mock()
+    mock_location.reset_mock()
+    mock_response.reset_mock()
```
Nitpick: You are creating all the mocks here, why do you need to reset them if they will probably be removed right after you reset them?

llucax reviewed (commented):
LGTM in general, just a few very minor comments.

llucax commented:
Not sure why GitHub is not allowing me to resolve conversations, maybe a permissions issue? I will check the permissions on this repo.

llucax commented:
Added the API and SDK team with write access to match other client repo permissions.

eduardiazf committed:
refactor(client): Update Microgrid serialization and remove deprecated functions

- Replaced the `to_dict` methods in `DeliveryArea` and `Location` classes with a new `to_json` method in the `Microgrid` class for improved JSON serialization.
- Removed the deprecated `delete_me` function from the codebase.
- Updated the `RELEASE_NOTES.md` to reflect these changes and added new required dependencies.

These modifications enhance the API's usability and maintainability while ensuring a cleaner codebase.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
fix(client): Update type hints in Microgrid class for better clarity

- Changed type hints for `delivery_area` and `location` attributes in the `Microgrid` class to support `DeliveryArea | None` and `Location | None`.
- This modification enhances type safety and improves code readability.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf commented:
> It looks like commit [`6902612` (#21)](https://github.com/frequenz-floss/frequenz-client-assets-python/pull/21/commits/690261201c8233595ae50933111dc14b595fca6d) is missing a title.

Done, I updated the commit message.

cwasicki commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,147 @@
+
+    latitude: float
+    longitude: float
+    country_code: str
```
The country code should be optional

cwasicki commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,147 @@
+            create_time=pb.create_timestamp.ToDatetime().replace(tzinfo=timezone.utc),
+        )
+
+    def to_json(self) -> str:
```
What is this for? Does it need to be part of the type?

eduardiazf replied:
we can move to cli because it is just using in the function `print_microgrid_details`

cwasicki commented on src/frequenz/client/assets/types.py:
```diff
@@ -0,0 +1,147 @@
+    def to_json(self) -> str:
+        """Convert the Microgrid instance to a JSON string."""
+        microgrid_dict = asdict(self)
+        microgrid_dict["id"] = int(self.id)
```
Is this future proof? I think @llucax just recently introduced a new type for this.

eduardiazf replied:
This is for prevent the following `TypeError`
<img width="508" height="47" alt="image" src="https://github.com/user-attachments/assets/b23bbeef-9f45-4b6f-8a76-be8483d6b68f" />

llucax replied:
You need to implement a [custom json serializer in this case](https://docs.python.org/3/library/json.html#basic-usage): `json.dump(asdict(the_dataclass), default=my_serializer)`. I agree it would be better to put this stuff in the CLI module, as it is only used for the CLI, but didn't want to add more comments to the commits because as @cwasicki said in the future this will be moved to common in the future so we'll have to address it anyway.

cwasicki commented on src/frequenz/client/assets/__main__.py:
```diff
@@ -0,0 +1,186 @@
+# License: MIT
```
In other clients we put this under a `cli` subfolder, I would suggest to do this here too. See e.g. https://github.com/frequenz-floss/frequenz-client-reporting-python/tree/v0.x.x/src/frequenz/client/reporting/cli

cwasicki commented on src/frequenz/client/assets/__main__.py:
```diff
@@ -0,0 +1,186 @@
+    export ASSETS_API_URL="grpc://api.example.com:5555"
+    export ASSETS_API_AUTH_KEY="your-key"
+    export ASSETS_API_SIGN_SECRET="your-secret"
+    python -m frequenz.client.assets microgrid 123
```
Have a look at the reporting repo for defining a dedicated script that is installed with the package: https://github.com/frequenz-floss/frequenz-client-reporting-python/blob/0309f57573ec11900665243bda3278fcd36b70d5/pyproject.toml#L40

eduardiazf committed:
fix(client): Update country_code type hint in Location class

- Changed the type hint for `country_code` in the `Location` class from `str` to `str | None` to allow for optional values.
- This modification enhances type safety and improves code clarity.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
feat(client): Add CLI entry point for Assets API client and remove deprecated module

- Introduced a new CLI entry point in `pyproject.toml` for the Assets API client, allowing users to interact with the API via command line.
- Removed the deprecated `__main__.py` file from the `assets` module, streamlining the codebase and enhancing maintainability.

These changes improve the usability of the Assets API client by providing a more accessible command-line interface.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
refactor(client): Simplify Microgrid JSON serialization in CLI

- Removed the deprecated `to_json` method from the `Microgrid` class.
- Updated the `print_microgrid_details` function to directly serialize the `Microgrid` instance using `asdict`, enhancing clarity and maintainability.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

cwasicki reviewed (commented):
LGTM, thanks a lot!

llucax commented:
LGTM too, the custom json serializer is only a nice-to-have for me.

eduardiazf committed:
chore: Update CODEOWNERS to include python-sdk-team as a default owner

- Added @frequenz-floss/python-sdk-team to the CODEOWNERS file alongside the existing api-assets-team, ensuring shared ownership and collaboration on repository changes.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax commented:
Feel free to force-merge as the codeowners file change won't be effective until it is merged.
